cmd/remote_enforcer.go

```
@@ -0,0 +1,279 @@
+
+		} else {
+			//Do nothing since we have added this flow already
+		}
+		if time.Since(starttime) > 2*time.Second {
+			//Send out everything we have in the payload
+			request.Payload = rpcpayload
+			err = s.Rpchdl.RemoteCall(statsContextID, "RPCSERVER.GetStats", &request, &response)
+
+			starttime = time.Now()
+		}
+
+	}
+}
+func (s *Server) connectStatsClient(statsClient *StatsClient) error {
+
+	statschannel := os.Getenv("STATSCHANNEL_PATH")
+	err := statsClient.Rpchdl.NewRPCClient(statsContextID, statschannel)
+	if err != nil {
+		log.WithFields(log.Fields{"package": "remote_enforcer", "error": err}).Error("Stats RPC client cannot connect")
+	}
+	_, err = statsClient.Rpchdl.GetRPCClient(statsContextID)
+
+	go statsClient.SendStats()
+	return err
+}
+
+//InitEnforcer exported
+func (s *Server) InitEnforcer(req rpcWrapper.Request, resp *rpcWrapper.Response) error {
+	collector := new(CollectorImpl)
+	collector.Flowentries = list.New()
+	s.Collector = collector
+	if !s.rpchdl.CheckValidity(&req) {
+		resp.Status = errors.New("Message Auth Failed")
+		return nil
+	}
+	payload := req.Payload.(rpcWrapper.InitRequestPayload)
+	usePKI := (payload.SecretType == tokens.PKIType)
+	//Need to revisit what is packet processor
+	if usePKI {
+		//PKI params
+		publicKeyAdder := tokens.NewPKISecrets(payload.PrivatePEM, payload.PublicPEM, payload.CAPEM, map[string]*ecdsa.PublicKey{})
+		s.Enforcer = enforcer.NewDefaultDatapathEnforcer(payload.ContextID, collector, nil, publicKeyAdder)
+	} else {
+		//PSK params
+		publicKeyAdder := tokens.NewPSKSecrets(payload.PublicPEM)
+		s.Enforcer = enforcer.NewDefaultDatapathEnforcer(payload.ContextID, collector, nil, publicKeyAdder)
+	}
+	s.Enforcer.Start()
+	statsClient := &StatsClient{collector: collector, s: s, FlowCache: cache.NewCacheWithExpiration(120*time.Second, 1000), Rpchdl: rpcWrapper.NewRPCWrapper()}
+	s.connectStatsClient(statsClient)
+
+	resp.Status = nil
+	return nil
```

This method should return an error, but you never return any err here